File: src/modes/custom/shellEnv.ts
import * as os from 'node:os';
import * as path from 'node:path';
import * as fs from 'node:fs';

// Safe, regex-only scanner for shell dotfiles. It NEVER sources or executes the
// file; it scans the entire file content for a known allow-list of Anthropic/SDK
// env vars and extracts them. Both `export KEY=...` and bare `KEY=...` forms are
// accepted, as are values inside an alias definition.
//
// Unknown keys are ignored, values are never logged, and the returned `source` is
// just the filename (e.g. '.zshrc') so the operator can see where the effective
// value came from.

export interface ResolveCustomEnvOptions {
  // Override the home directory used to locate dotfiles (for tests).
  homeDir?: string;
  // Inject file contents by basename, e.g. { '.zshrc': 'export ANTHROPIC_MODEL=...' }.
  // Files not present here are read from disk normally.
  files?: Record<string, string>;
}

export interface CustomEnvResult {
  env: Record<string, string>;
  hasDangerousFlag: boolean;
  source: string | undefined;
}

// Dotfiles scanned. Later files override earlier files; within a file, the last
// occurrence of a key wins.
const DEFAULT_FILES = ['.zshrc', '.zprofile', '.bashrc', '.bash_profile', '.bash_login', '.profile'];

// Allow-listed env vars only. Anything else in the file is ignored.
const ALLOWED_KEYS = new Set([
  'ANTHROPIC_BASE_URL',
  'ANTHROPIC_AUTH_TOKEN',
  'ANTHROPIC_API_KEY',
  'ANTHROPIC_MODEL',
  'ANTHROPIC_SMALL_FAST_MODEL',
  'API_TIMEOUT_MS',
]);

const DANGEROUS_FLAG = '--dangerously-skip-permissions';

export function resolveCustomEnv(opts: ResolveCustomEnvOptions = {}): CustomEnvResult {
  const homeDir = opts.homeDir ?? os.homedir();
  const injectedFiles = opts.files ?? {};

  let env: Record<string, string> = {};
  let source: string | undefined;
  let hasDangerousFlag = false;

  for (const file of DEFAULT_FILES) {
    let content: string | undefined;
    if (Object.prototype.hasOwnProperty.call(injectedFiles, file)) {
      content = injectedFiles[file];
    } else {
      try {
        content = fs.readFileSync(path.join(homeDir, file), 'utf-8');
      } catch {
        continue;
      }
    }
    if (content === undefined || content.length === 0) continue;

    if (content.includes(DANGEROUS_FLAG)) {
      hasDangerousFlag = true;
    }

    const extracted = extractEnv(content);
    if (extracted.sourceKeys.length > 0) {
      // Later files override earlier files; extractEnv already keeps the last
      // occurrence within the same file.
      env = { ...env, ...extracted.env };
      source = file;
    }
  }

  return { env, source, hasDangerousFlag };
}

// Extract ALLOWED_KEYS env assignments from the full file content. Supports:
//   export ANTHROPIC_MODEL="kimi-k2.7-code"
//   ANTHROPIC_BASE_URL='https://...'
//   API_TIMEOUT_MS=600000
//   alias kimi='ANTHROPIC_MODEL="kimi" claude'
// Within one file, the last assignment for a key wins.
function extractEnv(content: string): { env: Record<string, string>; sourceKeys: string[] } {
  const env: Record<string, string> = {};
  const sourceKeys: string[] = [];

  const keysPattern = [...ALLOWED_KEYS].map(escapeRegex).join('|');
  const pattern = new RegExp(
    `(?:^|\\s|;|')(?:export\\s+)?\\b(${keysPattern})\\s*=\\s*(?:"([^"]*)"|'([^']*)'|([^\\s;]*))`,
    'g',
  );

  let match: RegExpExecArray | null;
  while ((match = pattern.exec(content)) !== null) {
    const key = match[1];
    const rawValue = match[2] ?? match[3] ?? match[4] ?? '';
    if (ALLOWED_KEYS.has(key)) {
      env[key] = rawValue;
      if (!sourceKeys.includes(key)) sourceKeys.push(key);
    }
  }

  return { env, sourceKeys };
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

File: swift/Sources/DiscordAgentBridge/Provider/ShellEnv.swift
import Foundation

// Safe, regex-only scanner for shell dotfiles (TS `modes/custom/shellEnv.ts`).
// NEVER sources or executes the file; scans content for an allow-list of Anthropic/SDK
// env vars. Both `export KEY=...` and bare `KEY=...` forms are accepted, as are values
// inside an alias definition. Unknown keys ignored; values never logged.

/// Override home / inject file contents by basename for tests.
public struct ResolveCustomEnvOptions: Sendable, Equatable {
    public var homeDir: String?
    /// e.g. `{ ".zshrc": "export ANTHROPIC_MODEL=..." }`. Missing keys are read from disk.
    public var files: [String: String]?

    public init(homeDir: String? = nil, files: [String: String]? = nil) {
        self.homeDir = homeDir
        self.files = files
    }
}

public struct CustomEnvResult: Sendable, Equatable {
    public var env: [String: String]
    public var hasDangerousFlag: Bool
    public var source: String?

    public init(env: [String: String], hasDangerousFlag: Bool, source: String?) {
        self.env = env
        self.hasDangerousFlag = hasDangerousFlag
        self.source = source
    }
}

/// Dotfiles scanned. Later files override earlier; within a file, last occurrence wins.
private let defaultDotfiles = [
    ".zshrc", ".zprofile", ".bashrc", ".bash_profile", ".bash_login", ".profile",
]

/// Allow-listed env vars only.
private let allowedKeys: Set<String> = [
    "ANTHROPIC_BASE_URL",
    "ANTHROPIC_AUTH_TOKEN",
    "ANTHROPIC_API_KEY",
    "ANTHROPIC_MODEL",
    "ANTHROPIC_SMALL_FAST_MODEL",
    "API_TIMEOUT_MS",
]

private let dangerousFlag = "--dangerously-skip-permissions"

/// Extract allow-listed env vars from shell dotfiles under `homeDir` (or injected `files`).
public func resolveCustomEnv(_ opts: ResolveCustomEnvOptions = ResolveCustomEnvOptions()) -> CustomEnvResult {
    let homeDir = opts.homeDir ?? NSHomeDirectory()

    var env: [String: String] = [:]
    var source: String?
    var hasDangerousFlag = false

    for file in defaultDotfiles {
        let content: String?
        // Injected map wins (incl. empty string) so tests never touch real disk when key is set.
        if let files = opts.files, files.keys.contains(file) {
            content = files[file]
        } else {
            let path = (homeDir as NSString).appendingPathComponent(file)
            content = try? String(contentsOfFile: path, encoding: .utf8)
        }
        guard let content, !content.isEmpty else { continue }

        if content.contains(dangerousFlag) {
            hasDangerousFlag = true
        }

        let extracted = extractEnv(content)
        if !extracted.sourceKeys.isEmpty {
            for (k, v) in extracted.env { env[k] = v }
            source = file
        }
    }

    return CustomEnvResult(env: env, hasDangerousFlag: hasDangerousFlag, source: source)
}

/// Display label: `"Custom (model)"` when ANTHROPIC_MODEL is set, else `"Custom"`.
public func customBackendLabel(_ opts: ResolveCustomEnvOptions = ResolveCustomEnvOptions()) -> String {
    let env = resolveCustomEnv(opts).env
    if let model = env["ANTHROPIC_MODEL"], !model.isEmpty {
        // Operator-set env value: an SDK alias becomes its wire id, anything else prints as-is.
        return "Custom (\(modelDisplayText(model)))"
    }
    return "Custom"
}

// MARK: - private

private struct ExtractedEnv {
    var env: [String: String]
    var sourceKeys: [String]
}

/// Extract ALLOWED_KEYS assignments. Supports double/single/unquoted values and alias form.
/// Within one file, the last assignment for a key wins.
private func extractEnv(_ content: String) -> ExtractedEnv {
    var env: [String: String] = [:]
    var sourceKeys: [String] = []

    // TS: /(?:^|\s|;|')(?:export\s+)?\b(KEYS)\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s;]*))/g
    let keysAlt = allowedKeys.sorted().map(NSRegularExpression.escapedPattern(for:)).joined(separator: "|")
    let pattern =
        "(?:^|\\s|;|')(?:export\\s+)?\\b(\(keysAlt))\\s*=\\s*(?:\"([^\"]*)\"|'([^']*)'|([^\\s;]*))"
    guard let re = try? NSRegularExpression(pattern: pattern, options: []) else {
        return ExtractedEnv(env: [:], sourceKeys: [])
    }

    let ns = content as NSString
    let full = NSRange(location: 0, length: ns.length)
    re.enumerateMatches(in: content, options: [], range: full) { match, _, _ in
        guard let match, match.numberOfRanges >= 5 else { return }
        let key = ns.substring(with: match.range(at: 1))
        guard allowedKeys.contains(key) else { return }

        let rawValue: String
        if match.range(at: 2).location != NSNotFound {
            rawValue = ns.substring(with: match.range(at: 2))
        } else if match.range(at: 3).location != NSNotFound {
            rawValue = ns.substring(with: match.range(at: 3))
        } else if match.range(at: 4).location != NSNotFound {
            rawValue = ns.substring(with: match.range(at: 4))
        } else {
            rawValue = ""
        }

        env[key] = rawValue
        if !sourceKeys.contains(key) { sourceKeys.append(key) }
    }

    return ExtractedEnv(env: env, sourceKeys: sourceKeys)
}
